File: src/domain/workspace-locate.ts
import { Array as Arr, Option } from "effect"
import type { Workspace } from "../schema/workspace.js"

// ---------------------------------------------------------------------------
// workspace-locate — pure resolution of cwd / branch-query → Option<Workspace>
//
// Replaces the 5 copies of process.cwd() matching scattered across commands.
// No services, no IO.
// ---------------------------------------------------------------------------

/**
 * Locate a workspace by an optional branch query, otherwise by current
 * working directory.
 *
 * - Branch query precedence: exact → endsWith(`/${q}`) → includes(q).
 * - cwd matching: `cwd === path || cwd.startsWith(path + "/")` (path-prefix
 *   safe — `/a/bc` does not match a workspace rooted at `/a/b`).
 * - A branch query takes precedence over cwd: when provided, cwd is not
 *   consulted even if the branch query matches nothing.
 */
export const locateWorkspace = (
  workspaces: ReadonlyArray<Workspace>,
  query: { cwd: string; branch?: string }
): Option.Option<Workspace> => {
  if (query.branch !== undefined) {
    return locateByBranch(workspaces, query.branch)
  }
  return locateByCwd(workspaces, query.cwd)
}

const locateByBranch = (
  workspaces: ReadonlyArray<Workspace>,
  q: string
): Option.Option<Workspace> =>
  Arr.findFirst(workspaces, (w) => w.branch === q).pipe(
    Option.orElse(() => Arr.findFirst(workspaces, (w) => w.branch.endsWith(`/${q}`))),
    Option.orElse(() => Arr.findFirst(workspaces, (w) => w.branch.includes(q)))
  )

const locateByCwd = (
  workspaces: ReadonlyArray<Workspace>,
  cwd: string
): Option.Option<Workspace> =>
  Arr.findFirst(workspaces, (w) => cwd === w.path || cwd.startsWith(`${w.path}/`))

File: src/domain/workspace_locate.rs
use crate::schema::Workspace;

/// Locate a workspace by an optional branch query, otherwise by cwd.
///
/// - Branch query precedence: exact → ends_with("/{q}") → contains(q).
/// - cwd matching is path-prefix safe: `/a/bc` does not match a workspace
///   rooted at `/a/b`.
/// - A branch query takes precedence over cwd: when provided, cwd is not
///   consulted even if the query matches nothing.
pub fn locate_workspace<'a>(
    workspaces: &'a [Workspace],
    cwd: &str,
    branch: Option<&str>,
) -> Option<&'a Workspace> {
    if let Some(q) = branch {
        return locate_by_branch(workspaces, q);
    }
    workspaces
        .iter()
        .find(|w| cwd == w.path || cwd.starts_with(&format!("{}/", w.path)))
}

fn locate_by_branch<'a>(workspaces: &'a [Workspace], q: &str) -> Option<&'a Workspace> {
    workspaces
        .iter()
        .find(|w| w.branch == q)
        .or_else(|| {
            workspaces
                .iter()
                .find(|w| w.branch.ends_with(&format!("/{q}")))
        })
        .or_else(|| workspaces.iter().find(|w| w.branch.contains(q)))
}
